File: ApemanPlugin/src/main/kotlin/apeman_core/features_extraction/calculators/candidate/NumFieldAccessCandidateCalculator.java
package apeman_core.features_extraction.calculators.candidate;

import com.intellij.psi.*;
import org.jetbrains.research.groups.ml_methods.utils.ExtractionCandidate;

import java.util.ArrayList;

public class NumFieldAccessCandidateCalculator extends AbstractNumCandidateCalculator {

    public NumFieldAccessCandidateCalculator(ArrayList<ExtractionCandidate> candidates) {
        super(candidates);
    }

    @Override
    public JavaRecursiveElementVisitor createVisitor() {
        return new NumFieldAccessCandidateCalculator.Visitor();
    }

    private class Visitor extends CandidateVisitor {

        PsiClass currentClass = null;

        @Override
        public void visitClass(PsiClass aClass) {
            currentClass = aClass;
            super.visitClass(aClass);
            currentClass = null;
        }

        @Override
        public void visitReferenceElement(PsiJavaCodeReferenceElement reference) {
            super.visitReferenceElement(reference);
            PsiElement elem = reference.resolve();
            if (elem instanceof PsiField && isInsideMethod &&
                    ((PsiField) elem).getContainingClass() == currentClass) {
                updateCounters();
            }
        }
    }
}

File: ApemanPlugin/src/main/kotlin/apeman_core/features_extraction/calculators/candidate/NumFieldAccessCandidateCalculator.kt
package apeman_core.features_extraction.calculators.candidate

import apeman_core.base_entities.ExtractionCandidate
import apeman_core.base_entities.FeatureType
import com.intellij.psi.*

class NumFieldAccessCandidateCalculator(candidates: List<ExtractionCandidate>
) : AbstractNumCandidateCalculator(candidates, FeatureType.NUM_FIELD_ACCESS) {

    override fun createVisitor() = Visitor()

    inner class Visitor : AbstractNumCandidateCalculator.CandidateVisitor() {

        private var currentClass: PsiClass? = null

        override fun visitClass(aClass: PsiClass) {
            currentClass = aClass
            super.visitClass(aClass)
            currentClass = null
        }

        override fun visitReferenceElement(reference: PsiJavaCodeReferenceElement) {
            super.visitReferenceElement(reference)
            val elem = reference.resolve()

            if (elem is PsiField && isInsideMethod) {
                updateCounters()
            }
        }
    }
}
